test/test_hmc_util.py

```diff
@@ -0,0 +1,366 @@
+        q_f, p_f, _, _ = lax.fori_loop(0, num_steps,
+                                       lambda i, val: vv_update(step_size, val),
+                                       vv_state)
+        return (q_f, p_f)
+
+    model, args = example
+    fn = jit(get_final_state, static_argnums=(0,)) if jitted else get_final_state
+    q_f, p_f = fn(model, args.step_size, args.num_steps, args.q_i, args.p_i)
+
+    logger.info('Test trajectory:')
+    logger.info('initial q: {}'.format(args.q_i))
+    logger.info('final q: {}'.format(q_f))
+    for node in args.q_f:
+        assert_allclose(q_f[node], args.q_f[node], atol=args.prec)
+        assert_allclose(p_f[node], args.p_f[node], atol=args.prec)
+
+    logger.info('Test energy conservation:')
+    energy_initial = model.kinetic_fn(args.p_i) + model.potential_fn(args.q_i)
+    energy_final = model.kinetic_fn(p_f) + model.potential_fn(q_f)
+    logger.info('initial energy: {}'.format(energy_initial))
+    logger.info('final energy: {}'.format(energy_final))
+    assert_allclose(energy_initial, energy_final, atol=1e-5)
+
+    logger.info('Test time reversibility:')
+    p_reverse = tree_map(lambda x: -x, p_f)
+    q_i, p_i = get_final_state(model, args.step_size, args.num_steps, q_f, p_reverse)
+    for node in args.q_i:
+        assert_allclose(q_i[node], args.q_i[node], atol=1e-4)
+
+
+@pytest.mark.parametrize('jitted', [True, False])
+@pytest.mark.parametrize('init_step_size', [0.1, 10.0])
+def test_find_reasonable_step_size(jitted, init_step_size):
+    def kinetic_fn(p):
+        return 0.5 * p ** 2
+
+    def potential_fn(q):
+        return 0.5 * q ** 2
+
+    p_generator = lambda: 1.0  # noqa: E731
+    q = 0.0
+
+    fn = (jit(find_reasonable_step_size, static_argnums=(0, 1, 2))
+          if jitted else find_reasonable_step_size)
+    step_size = fn(potential_fn, kinetic_fn, p_generator, q, init_step_size)
+
+    # Apply 1 velocity verlet step with step_size=eps, we have
+    # z_new = eps, r_new = 1 - eps^2 / 2, hence energy_new = 0.5 + eps^4 / 8,
+    # hence delta_energy = energy_new - energy_init = eps^4 / 8.
+    # We want to find a reasonable step_size such that delta_energy ~ -log(0.8),
+    # hence that step_size ~ the following threshold
+    threshold = np.power(-np.log(0.8) * 8, 0.25)
+
+    # Confirm that given init_step_size, we will doubly increase/decrease it
+    # until it passes threshold.
+    if init_step_size < threshold:
+        assert step_size / 2 < threshold
+        assert step_size > threshold
+    else:
+        assert step_size * 2 > threshold
+        assert step_size < threshold
+
+
+@pytest.mark.parametrize("num_steps, expected", [
+    (18, [(0, 17)]),
+    (50, [(0, 6), (7, 44), (45, 49)]),
+    (100, [(0, 14), (15, 89), (90, 99)]),
+    (150, [(0, 74), (75, 99), (100, 149)]),
+    (200, [(0, 74), (75, 99), (100, 149), (150, 199)]),
+    (280, [(0, 74), (75, 99), (100, 229), (230, 279)]),
+])
+def test_build_adaptation_schedule(num_steps, expected):
```

Comment:
I thought we already had these tests? Should we remove the older tests.

Reply:
@neerajprad I just rename the files so no new tests are added.

Reply:
Oppss. Wait, I need to separate the tests from distribution utils too.

Reply:
done!